src/connector/src/source/filesystem/opendal_source/opendal_reader.rs

```diff
@@ -0,0 +1,172 @@
+{
+    #[try_stream(boxed, ok = StreamChunkWithState, error = RwError)]
+    async fn into_chunk_stream(self) {
+        for split in self.splits {
+            let actor_id = self.source_ctx.source_info.actor_id.to_string();
+            let source_id = self.source_ctx.source_info.source_id.to_string();
+            let source_ctx = self.source_ctx.clone();
+
+            let split_id = split.id();
+
+            let data_stream =
+                Self::stream_read_object(self.connector.op.clone(), split, self.source_ctx.clone());
+
+            let parser =
+                ByteStreamSourceParserImpl::create(self.parser_config.clone(), source_ctx).await?;
+            let msg_stream = if matches!(
+                parser,
+                ByteStreamSourceParserImpl::Json(_) | ByteStreamSourceParserImpl::Csv(_)
+            ) {
+                parser.into_stream(nd_streaming::split_stream(data_stream))
+            } else {
+                parser.into_stream(data_stream)
+            };
+            #[for_await]
+            for msg in msg_stream {
+                let msg = msg?;
+                self.source_ctx
+                    .metrics
+                    .partition_input_count
+                    .with_label_values(&[&actor_id, &source_id, &split_id])
+                    .inc_by(msg.chunk.cardinality() as u64);
+                yield msg;
+            }
+        }
+    }
+
+    #[try_stream(boxed, ok = Vec<SourceMessage>, error = anyhow::Error)]
+    pub async fn stream_read_object(
+        op: Operator,
+        split: OpendalFsSplit<C>,
+        source_ctx: SourceContextRef,
+    ) {
+        let actor_id = source_ctx.source_info.actor_id.to_string();
+        let source_id = source_ctx.source_info.source_id.to_string();
+        let max_chunk_size = source_ctx.source_ctrl_opts.chunk_size;
+        let split_id = split.id();
+
+        let object_name = split.name.clone();
+
+        let reader = op
+            .reader_with(&object_name)
+            .range(split.offset as u64..)
+            .await?;
+
+        let stream_reader = StreamReader::new(
+            reader.map_err(|e| std::io::Error::new(std::io::ErrorKind::Other, e)),
+        );
+        let buf_reader = BufReader::new(stream_reader);
+        let stream = ReaderStream::with_capacity(buf_reader, STREAM_READER_CAPACITY);
+
+        let mut offset: usize = split.offset;
+        let mut batch_size: usize = 0;
+        let mut batch = Vec::new();
+        #[for_await]
+        for read in stream {
+            let bytes = read?;
+            let len = bytes.len();
+            let msg = SourceMessage {
+                key: None,
+                payload: Some(bytes.as_ref().to_vec()),
+                offset: offset.to_string(),
+                split_id: split.id(),
+                meta: SourceMeta::Empty,
+            };
+            offset += len;
+            batch_size += len;
+            batch.push(msg);
+            if batch.len() >= max_chunk_size {
```

Comment:
Just want to make sure we won't have unbounded buffer. Is the size of `msg` fixed or bounded (maybe by `STREAM_READER_CAPACITY=4KB`)? With the default `max_chunk_size=256`, without considering the delimter, does it mean that we will never emit a chunk >= 1MB?

Reply:
> Just want to make sure we won't have unbounded buffer

Just realized yesterday, RW can be vulnerable in this case.

Reply:
> > Just realized yesterday, RW can be vulnerable in this case.
> 
> Do you mean that the msg returned in ReaderStream can be arbitarily large?

parser holds a buffer to handle truncated files. But if there are no delimiter found, the buffer can be really large.  Let's fix this in another pr